diff --git a/truffle/mx.truffle/suite.py b/truffle/mx.truffle/suite.py
index 6948225d6bd..ecf970ce7c2 100644
--- a/truffle/mx.truffle/suite.py
+++ b/truffle/mx.truffle/suite.py
@@ -1943,18 +1943,8 @@ suite = {
           "windows-amd64",
           "windows-aarch64",
       ],
-      "os": {
-        "linux": {
-          "layout": {
-            # only glibc.
-            "META-INF/resources/nfi-native/libnfi/<os>/<arch>/bin/" : "dependency:com.oracle.truffle.nfi.native/linux-*/glibc/*"
-          },
-        },
-        "<others>": {
-          "layout": {
-            "META-INF/resources/nfi-native/libnfi/<os>/<arch>/bin/" : "dependency:com.oracle.truffle.nfi.native/*/*/*",
-          },
-        },
+      "layout": {
+        "META-INF/resources/nfi-native/libnfi/<os>/<arch>/bin/" : "dependency:com.oracle.truffle.nfi.native/*/*/*",
       },
       "description" : "Contains the native library needed by the libffi NFI backend.",
       "maven": False,
@@ -2300,18 +2290,8 @@ suite = {
       "native" : True,
       "platformDependent" : True,
       "description" : "Truffle NFI support distribution for the GraalVM",
-      "os": {
-        "linux": {
-          "layout": {
-            # only glibc.
-            "./" : "dependency:com.oracle.truffle.nfi.native/linux-*/glibc/*"
-          },
-        },
-        "<others>": {
-          "layout": {
-            "./" : "dependency:com.oracle.truffle.nfi.native/*/*/*",
-          },
-        },
+      "layout": {
+        "./" : "dependency:com.oracle.truffle.nfi.native/*/*/*",
       },
       "maven" : False,
       "graalCompilerSourceEdition": "ignore",
